alg/viewshed/combiner.cpp

```diff
@@ -0,0 +1,78 @@
+/******************************************************************************
+ * (c) 2024 info@hobu.co
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#include "combiner.h"
+#include "util.h"
+
+namespace gdal
+{
+namespace viewshed
+{
+
+/// Read viewshed executor output and sum it up in our owned memory raster.
+void Combiner::run()
+{
+    DatasetPtr pTempDataset;
+
+    while (m_inputQueue.pop(pTempDataset))
+    {
+        if (!m_dataset)
+            m_dataset = std::move(pTempDataset);
+        else
+            sum(std::move(pTempDataset));
+    }
+    // Queue remaining summed rasters.
+    queueOutputBuffer();
+}
+
+/// Add the values of the source dataset to those of the owned dataset.
+/// @param src  Source dataset.
+void Combiner::sum(DatasetPtr src)
+{
+    if (!m_dataset)
+    {
+        m_dataset = std::move(src);
+        return;
+    }
+    size_t size = bandSize(*m_dataset->GetRasterBand(1));
+
+    uint8_t *dstP =
+        static_cast<uint8_t *>(m_dataset->GetInternalHandle("MEMORY1"));
+    uint8_t *srcP = static_cast<uint8_t *>(src->GetInternalHandle("MEMORY1"));
+    for (size_t i = 0; i < size; ++i)
+        *dstP++ += *srcP++;
```

Comment:
Isn't there a risk of overflowing uint8_t ? If so, maybe:
```suggestion
    {
        *dstP = std::min<uint8_t>(*dstP +*srcP, 255);
        ++dstP;
        ++srcP;
    }
```

Reply:
hum, I assume my above comment might be useless, given below comment ?

Reply:
Yes. When we've done this 255 times, we move the buffer off for later processing and start over.